### ISSUE TITLE
Numba changes behavior of np.flip when assigning flipped output back into input array

### ISSUE THREAD

**Issue by acorver** [numba/numba#8213]
## Reporting a bug

While debugging a problem with a larger computation, I traced the issue to a change in behavior of **numpy.flip** once the **@numba.njit()** decorator is added.

Here is the reproducible code:

    import numpy as np, numba as nb
    
    print('NUMBA VERSION: {}'.format(nb.__version__))
    print('NUMPY VERSION: {}'.format(np.__version__))
    
    @nb.njit()
    def fnNumba2d(numy, offsetx):
        x = np.arange(numy, dtype=np.int64).reshape((1, numy))
        print('INTERMEDIATE RESULT FOR 2D NP.FLIP AT OFFSET ', offsetx, '    (NUMBA): ', np.flip(x[0, offsetx:(offsetx+2)]))
        x[0, offsetx:(offsetx+2)] = np.flip(x[0, offsetx:(offsetx+2)])
        print('FINAL        RESULT FOR 2D NP.FLIP AT OFFSET ', offsetx, '    (NUMBA): ', x[0, offsetx:(offsetx+2)])
        return x
    
    def fnNoNumba2d(numy, offsetx):
        x = np.arange(numy, dtype=np.int64).reshape((1, numy))
        print('INTERMEDIATE RESULT FOR 2D NP.FLIP AT OFFSET ', offsetx, ' (NO NUMBA): ', np.flip(x[0, offsetx:(offsetx+2)]))
        x[0, offsetx:(offsetx+2)] = np.flip(x[0, offsetx:(offsetx+2)])
        print('FINAL        RESULT FOR 2D NP.FLIP AT OFFSET ', offsetx, ' (NO NUMBA): ', x[0, offsetx:(offsetx+2)])
        return x
    
    @nb.njit()
    def fnNumba1d(numy, offsetx):
        x = np.arange(numy, dtype=np.int64)
        x[offsetx:(offsetx+2)] = np.flip(x[offsetx:(offsetx+2)])
        return x
    
    def fnNoNumba1d(numy, offsetx):
        x = np.arange(numy, dtype=np.int64)
        x[offsetx:(offsetx+2)] = np.flip(x[offsetx:(offsetx+2)])
        return x
    
    # NO DIFFERENCE
    np.testing.assert_array_equal(fnNoNumba1d(3, 0), fnNumba1d(3, 0), err_msg="FAILED: FLIP ELEMENTS 0<->1 IN ARRAY SIZE=3")
    
    # NO DIFFERENCE
    np.testing.assert_array_equal(fnNoNumba1d(3, 1), fnNumba1d(3, 1), err_msg="FAILED: FLIP ELEMENTS 1<->2 IN ARRAY SIZE=3")
    
    # NO DIFFERENCE
    np.testing.assert_array_equal(fnNoNumba2d(3, 0), fnNumba2d(3, 0), err_msg="FAILED: FLIP ELEMENTS 0<->1 IN ARRAY SIZE=1x3")
    
    # NUMBA OUTPUT DIFFERS FROM PYTHON
    np.testing.assert_array_equal(fnNoNumba2d(3, 1), fnNumba2d(3, 1), err_msg="FAILED: FLIP ELEMENTS 1<->2 IN ARRAY SIZE=1x3")

This produces the following output:

    NUMBA VERSION: 0.56.0rc1+0.gea8407264.dirty
    NUMPY VERSION: 1.22.4
    INTERMEDIATE RESULT FOR 2D NP.FLIP AT OFFSET  0  (NO NUMBA):  [1 0]
    FINAL        RESULT FOR 2D NP.FLIP AT OFFSET  0  (NO NUMBA):  [1 0]
    INTERMEDIATE RESULT FOR 2D NP.FLIP AT OFFSET  0     (NUMBA):  [1 0]
    FINAL        RESULT FOR 2D NP.FLIP AT OFFSET  0     (NUMBA):  [1 0]
    INTERMEDIATE RESULT FOR 2D NP.FLIP AT OFFSET  1  (NO NUMBA):  [2 1]
    FINAL        RESULT FOR 2D NP.FLIP AT OFFSET  1  (NO NUMBA):  [2 1]
    INTERMEDIATE RESULT FOR 2D NP.FLIP AT OFFSET  1     (NUMBA):  [2 1]
    FINAL        RESULT FOR 2D NP.FLIP AT OFFSET  1     (NUMBA):  [2 2]
    
    ---------------------------------------------------------------------------
    AssertionError                            Traceback (most recent call last)
    Input In [16], in <cell line: 42>()
         39 np.testing.assert_array_equal(fnNoNumba2d(3, 0), fnNumba2d(3, 0), err_msg="FAILED: FLIP ELEMENTS 0<->1 IN ARRAY SIZE=1x3")
         41 # NUMBA OUTPUT DIFFERS FROM PYTHON
    ---> 42 np.testing.assert_array_equal(fnNoNumba2d(3, 1), fnNumba2d(3, 1), err_msg="FAILED: FLIP ELEMENTS 1<->2 IN ARRAY SIZE=1x3")
    
        [... skipping hidden 1 frame]
    
    File ~\.conda\envs\numbabug\lib\site-packages\numpy\testing\_private\utils.py:844, in assert_array_compare(comparison, x, y, err_msg, verbose, header, precision, equal_nan, equal_inf)
        840         err_msg += '\n' + '\n'.join(remarks)
        841         msg = build_err_msg([ox, oy], err_msg,
        842                             verbose=verbose, header=header,
        843                             names=('x', 'y'), precision=precision)
    --> 844         raise AssertionError(msg)
        845 except ValueError:
        846     import traceback
    
    AssertionError: 
    Arrays are not equal
    FAILED: FLIP ELEMENTS 1<->2 IN ARRAY SIZE=1x3
    Mismatched elements: 1 / 3 (33.3%)
    Max absolute difference: 1
    Max relative difference: 0.5
     x: array([[0, 2, 1]], dtype=int64)
     y: array([[0, 2, 2]], dtype=int64)
    
To summarize: If a submatrix within a 2D array is flipped and assigned back into the input, and if this submatrix is not located at column index 0, entries are duplicated rather than flipped.

Is this indeed a bug, or is it not recommended to assign the output of **numpy.flip** back into the input array? (Given that **numpy.flip** returns a view rather than a copy.)
